src/Dfc.CourseDirectory.Functions/FixVenues/VenueAnalyser.cs

```diff
@@ -0,0 +1,103 @@
+                {
+                    Apprenticeship = apprenticeship,
+                    UnfixableVenueReason = UnfixableApprenticeshipVenueReasons.ProviderHasNoLiveVenues
+                };
+            }
+
+            return new ApprenticeshipVenueCorrection
+            {
+                Apprenticeship = apprenticeship,
+                ApprenticeshipLocationVenueCorrections = RelevantApprenticeshipLocations(apprenticeship.ApprenticeshipLocations)
+                    .Select(location => AnalyseLocation(location, providerVenues))
+                    .Where(apprenticeshipLocationVenueCorrection => apprenticeshipLocationVenueCorrection != null)
+                    .ToList()
+            };
+        }
+
+        private ApprenticeshipLocationVenueCorrection AnalyseLocation(ApprenticeshipLocation location, IReadOnlyCollection<Venue> providersVenues)
+        {
+            CorruptionType corruptionType;
+            if (providersVenues.Any(v => v.Id == location.VenueId))
+            {
+                return null; // valid record, nothing to do
+            }
+
+            if (location.VenueId == null)
+            {
+                corruptionType = CorruptionType.NullVenueId;
+            }
+            else if (location.VenueId == Guid.Empty)
+            {
+                corruptionType = CorruptionType.EmptyVenueId;
+            }
+            else
+            {
+                corruptionType = CorruptionType.VenueNotInProvidersLiveVenueList;
+            }
+
+            var matchingVenues = _venueCorrectionFinder.GetMatchingVenues(location, providersVenues);
+
+            UnfixableLocationVenueReasons? unfixableLocationVenueReason = null;
+            Venue matchingVenue = null;
+            IReadOnlyList<Venue> duplicateVenues = null;
+            if (matchingVenues == null || !matchingVenues.Any())
+            {
+                unfixableLocationVenueReason = UnfixableLocationVenueReasons.NoMatchingVenue;
+            }
+            else if (matchingVenues.Count() > 1)
+            {
+                unfixableLocationVenueReason = UnfixableLocationVenueReasons.DuplicateMatchingVenues;
+                duplicateVenues = matchingVenues;
+            }
+            else
+            {
+                matchingVenue = matchingVenues.Single();
+            }
+
+            return new ApprenticeshipLocationVenueCorrection
+            {
+                LocationId = location.Id,
+                VenueIdOriginal = location.VenueId,
+                CorruptionType = corruptionType,
+                VenueCorrection = matchingVenue,
+                UnfixableLocationVenueReason = unfixableLocationVenueReason,
+                DuplicateVenues = duplicateVenues,
+            };
+        }
+
+        /// <summary>
+        /// Find locations that are worth fixing
+        /// </summary>
+        public static IEnumerable<ApprenticeshipLocation> RelevantApprenticeshipLocations(IEnumerable<ApprenticeshipLocation> locations)
```

Comment:
Shouldn't this be `private`?

Reply:
it's being tested directly here https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/71807102ad94e01544da74f1a8e2d827b5c2633a/tests/Dfc.CourseDirectory.Functions.Tests/VenueFixup/VenueAnalyserTests.cs#L251

better idea welcome, I guess I could spin out a class but didn't seem worth it on this one

Reply:
The approach is a smell (sic. _Single Responsibility_), but given that this is throwaway code and it's confined to `Functions.FixVenues` I don't see an issue with leaving it as is.